cpp/src/arrow/compute/kernels/scalar_validity.cc

```diff
@@ -37,23 +37,22 @@ struct IsValidOperator {
   static void Call(KernelContext* ctx, const ArrayData& arr, ArrayData* out) {
     DCHECK_EQ(out->offset, 0);
     DCHECK_LE(out->length, arr.length);
-    if (arr.buffers[0] != nullptr) {
-      out->buffers[1] = arr.offset == 0
-                            ? arr.buffers[0]
-                            : SliceBuffer(arr.buffers[0], arr.offset / 8, arr.length / 8);
+    if (arr.null_count != 0 && arr.buffers[0] != nullptr) {
```

Comment:
Note: I added a helper function `ArrayData::MayHaveNulls` to do this since it's such a common pattern